src/tools/perf/lib/ucp_tests.cc

```diff
@@ -301,6 +297,22 @@ class ucp_perf_test_runner {
         }
     }
 
+    inline void wait_for_sn(void *buffer, size_t length, uint8_t sn)
+    {
+        volatile uint8_t *ptr = (volatile uint8_t*)buffer + length - 1;
+        while (*ptr != sn) {
+            progress_responder();
+        }
+    }
+
+    void wait_for_last_sn(void *buffer, size_t length, uint8_t last_sn)
+    {
+        if ((CMD == UCX_PERF_CMD_PUT) &&
+            (TYPE == UCX_PERF_TEST_TYPE_STREAM_UNI)) {
+            wait_for_sn(buffer, length, last_sn);
```

Comment:
how does it work with SN wraparound?

Reply:
we do 100 `ucp_put` consequentially, so it's guaranteed that we will see the last data (i.e. the last `sn`) that comes to the responder, since it won't be wraparound by the previous data

Reply:
but can we exit too soon from this loop?

Reply:
> but can we exit too soon from this loop?

I guess it has to be guaranteed by sw/hw that data comes in order, so, when shows that the last byte is the last `sn`, so it means that all data was already received.
the error with the loop waiting for the concrete `sn` (0, 1, 2, ... <last sn>) appeared since the responder wasn't quick enoug to read e.g. `sn=10` that 's expected, when `sn=11` is already written to the receive buffer